File: ts/game.ts
const LIVE = "\u25FD";
const DEAD = "\u25FE";
const CLEAR = "\x1b[1A\x1b[0K";

class World {
  readonly dim: number;
  state: boolean[];

  constructor(dim: number) {
    this.dim = dim;
    this.state = new Array(dim * dim);
  }

  randomize() {
    for (let i = 0; i < this.state.length; i++) {
      this.state[i] = Math.random() < 0.5;
    }
  }

  get(i: number, j: number): boolean {
    return this.state[this.index(i, j)]!;
  }

  set(i: number, j: number, value: boolean) {
    this.state[this.index(i, j)] = value;
  }

  liveNeighbours(i: number, j: number) {
    return [
      this.get(i - 1, j - 1),
      this.get(i - 1, j),
      this.get(i - 1, j + 1),
      this.get(i, j - 1),
      this.get(i, j + 1),
      this.get(i + 1, j - 1),
      this.get(i + 1, j),
      this.get(i + 1, j + 1),
    ].reduce((acc, live) => (live ? acc + 1 : acc), 0);
  }

  private index(i: number, j: number) {
    i = (i + this.dim) % this.dim;
    j = (j + this.dim) % this.dim;
    return i * this.dim + j;
  }
}

class Game {
  world: World;

  constructor(dim: number) {
    this.world = new World(dim);
    this.world.randomize();
  }

  async run() {
    await this.flush();

    let running = true;
    let rendered = false;
    process.on("SIGINT", () => {
      running = false;
    });

    while (running) {
      await this.render(rendered);
      rendered = true;

      this.next();
      await this.sleep(200);
    }

    process.stdout.write("\rBye");
  }

  next() {
    const newWorld = new World(this.world.dim);
    for (let i = 0; i < this.world.dim; i++) {
      for (let j = 0; j < this.world.dim; j++) {
        let cell = this.world.get(i, j);
        const liveNeighbours = this.world.liveNeighbours(i, j);
        if (cell) {
          cell = liveNeighbours === 2 || liveNeighbours === 3;
        } else {
          cell = liveNeighbours === 3;
        }
        newWorld.set(i, j, cell);
      }
    }
    this.world = newWorld;
  }

  async render(rendered: boolean) {
    if (rendered) {
      process.stdout.write(CLEAR.repeat(this.world.dim));
    }

    for (let i = 0; i < this.world.dim; i++) {
      for (let j = 0; j < this.world.dim; j++) {
        const cell = this.world.get(i, j);
        process.stdout.write(cell ? LIVE : DEAD);
      }
      process.stdout.write("\n");
    }

    await this.flush();
  }

  private sleep(ms: number) {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }

  private flush() {
    return new Promise((resolve) => process.stdout.write("", resolve));
  }
}

const game = new Game(32);
await game.run();

File: julia/game.jl
using Random: bitrand
Base.exit_on_sigint(false)

const DIM = 32

const LIVE = "\u25FD"
const DEAD = "\u25FE"
const CLEAR = "\e[1A\e[K" ^ DIM

const SLEEP_S = 0.2

function run()
    state = bitrand(DIM, DIM)
    render(state)
    while true
        sleep(SLEEP_S)
        state = next(state)
        print(CLEAR)
        render(state)
    end
end

function render(state::BitMatrix)
    for i=1:DIM
        for j=1:DIM
            print(state[i, j] ? LIVE : DEAD)
        end
        println()
    end
end

function next(state::BitMatrix)::BitMatrix
    BitMatrix(survives(state, i, j) for i=1:DIM, j=1:DIM)
end

function survives(state::BitMatrix, i::Int, j::Int)
    nx = mod1.([
        i-1  j-1
        i-1  j    
        i-1  j+1
        i    j-1
        i    j+1
        i+1  j-1
        i+1  j    
        i+1  j+1
    ], DIM)

    live_neighbours = count(state[ix, jx] for (ix, jx) in eachrow(nx))

    if state[i, j]
        live_neighbours == 2 || live_neighbours == 3
    else
        live_neighbours == 3
    end
end

try
    run()
catch e
    if e isa InterruptException
        println("Bye")
    else
        rethrow(e)
    end
end
